### DISCUSSION TITLE
XDG specs: implement ourselves or add a dependency?

### DISCUSSION THREAD

**Discussion by sheybey** [Chatterino/chatterino2#4366] [Q&A] [answered]
I'm planning to implement support for opening links in a private window for linux (read: freedesktop-compliant DE) users. This involves a few freedesktop specs:

- Shared MIME database: https://www.freedesktop.org/wiki/Specifications/shared-mime-info-spec
- Default apps: https://www.freedesktop.org/wiki/Specifications/mime-apps-spec/
- Desktop entries: https://www.freedesktop.org/wiki/Specifications/desktop-entry-spec/
- Base directories: https://www.freedesktop.org/wiki/Specifications/basedir-spec/

These are all simple enough on their own that I wouldn't worry too much about writing a bespoke implementation of one of them for chatterino. However, the way that chatterino implements this feature (look up the user's default browser and construct its own command line) involves enough specifications that I think it's worth discussing adding a dependency.

That being said, there are a number of apps that use their own implementations (see the shared mime spec for a list) and QT itself implements a small subset of the base directories and desktop entries spec (desktop entries are just INI with well known keys, so you can read them with `QSettings`.) So, implementing them in chatterino is not entirely unreasonable.

As for a dependency, the simple reference implementation (https://gitlab.freedesktop.org/xdg/xdgmime) looks like it would suffice, with the caveat that the README implies it's meant to be vendored. Anecdotally, it doesn't appear to be in debian's repositories, either.

Another point this brings up, perhaps out of the scope of this discussion, is that all of the work of finding the user's browser is done on the GUI thread currently. The windows implementation is a couple COM calls, which is fast enough IMO. The freedesktop implementation would be a program invocation of `xdg-settings` if the DE provides it, and if it fails opening 2 files (mime database and browser desktop entry,) searching several directories for each. This is approaching enough activity that it might be worth putting it on the event loop so the interface doesn't freeze up.

**Comment by pajlada** [comment, marked as answer]
For the performance concern, I don't think it's something we need for a V1 as it would still be something that's done on a user action, not randomly.

If possible, an as-small-as-possible-implementation written in Chatterino seems like an ok approach rather than adding a vendored dependency.
